Set up Gradle Projects for multi-platform foundation (common, fabric, neoforge)

EchoEllet opened the pull request:
Setups `common`, `neoforge`, and `fabric` Gradle projects (AKA modules) for multi-platform progress, without using third-party templates, plugins, or dependencies (i.e., **without Architectury API or Gradle Plugin**).

The source code has been moved fully from the root project `src/` to `neoforge/src/`, while the vanilla assets (including `assets/epicfight`) have been moved to `common` but some NeoForge exclusive assets, such as `data/neoforge` kept under `neoforge`, for clear separation of concerns, so they are not included in the Fabric JAR file.

> [!TIP]
> This is **100% backward compatible with existing addons**, since it does not affect how addons import the Java classes.
> A single Gradle project may contain multiple packages and resources. All packages and resources under `common` are included in `neoforge` and `fabric`, but not vice versa.

> [!NOTE]
> This was done using standard Gradle projects, promoting best practices whenever possible, such as using Gradle convention plugins (`gradle/build-logic`).
> `gradle/build-logic` (precompiled plugins) is newer than `buildSrc`, for faster builds, and is officially recommended by Gradle, while supporting Gradle 9, configuration-cache, and up-to-date tools.

<details><summary>The Fabric version of Epic Fight is empty.</summary>
<p>

Can be run via `./gradlew fabric:runClient`.

<img width="843" height="505" alt="image" src="https://github.com/user-attachments/assets/4348280e-58aa-4295-a153-d007313b1c15" />

</p>
</details> 

> [!IMPORTANT]
> It's important to use `./gradlew neoforge:publishMods` and not `./gradlew publishMods`, otherwise the empty Fabric version will be published to Modrinth and CurseForge, which is only appropriate in April.

<details><summary>Publish the NeoForge version using the Gradle plugin via the IntelliJ IDEA GUI</summary>
<p>

<img width="890" height="1254" alt="image" src="https://github.com/user-attachments/assets/153fdc5a-9c13-4e2d-8473-eb8797d8f84e" />

</p>
</details> 

<details><summary>Screenshots while crafting</summary>
<p>

<img width="3230" height="2158" alt="image" src="https://github.com/user-attachments/assets/2a12c12d-484e-4b3a-bdb7-711ac7d22006" />


</p>
</details> 

<details><summary>NeoForge Vanilla-mode</summary>
<p>

NeoForge provides [NeoForm](https://github.com/neoforged/NeoForm) ([vanilla mode](https://github.com/neoforged/ModDevGradle#vanilla-mode)), which allows using the NeoForge toolchain in the `common` project, but vanilla Minecraft classes, without any patches or NeoForge APIs.

<img width="2102" height="1928" alt="image" src="https://github.com/user-attachments/assets/6117e4ac-43eb-49c1-9cb3-7c22de037b05" />

This is used in the `common` project, so even though there is the `neoforge {}` block, it's not NeoForge, and this was made for compatibility reasons.
This approach is friendly with other mod-platform such as Fabric.

</p>
</details> 

### Motivation

<details><summary>Why was this done?</summary>
<p>

We can't progress without this Gradle setup, since it's more than just abstracting `net.neoforged.*`.

Forge-like platforms patch vanilla classes to add new fields, methods, and constructors.

> **Note:** NeoForge is improving the situation in MC 1.21.9+ and 26+.

The same vanilla class may not work outside of Forge and will be silently dependent on Forge

<img width="1846" height="1024" alt="image" src="https://github.com/user-attachments/assets/0e334e6f-f377-427c-a68d-90d17e5d7303" />

<img width="1966" height="1140" alt="image" src="https://github.com/user-attachments/assets/e1876817-a6e3-4313-a0f4-c2872c5b3e1f" />

For example, `Mob#spawnType` and `Mob#spawnCancelled` were added by Forge, and it's hard to tell whether a specific MC API was added by Forge or not without looking at this directory: https://github.com/neoforged/NeoForge/tree/1.21.x/patches/net/minecraft (NeoForge patches).

The ideal solution is to set up Gradle projects (AKA modules), which are `common`, `fabric`, and `neoforge`,
where the Java source code is fully in `neoforge` (`src/main` -> `neoforge/src/main`).

The situation can be improved by slowly abstracting and moving classes to `common` with the corresponding `neoforge` and `fabric` platform implementations to catch early
design issues.

This is not **a code-breaking change**; for example, if a class was moved from `neoforge` to `common` without any source modification, addons can still import the same class with no behavior changes.

</p>
</details> 

<details><summary>Note regarding existing and new code</summary>
<p>

1. New features should be in `common`, designed in mind to be independent of a specific mod-loader.
2. Existing features can be slowly moved from `neoforge` to `common`, and then implement platform-specific methods in `fabric`.

</p>
</details> 

<details><summary>Isn't this maintenance burden?</summary>
<p>

A multi-platform architecture does not exists primarly to support Fabric, but to be independent of Forge or a specific mod-loader, and therefore a more future-proof project.

This will be significantly useful if:

1. Minecraft introduced official modding APIs.
2. Forge/NeoForge got forked again.
3. The maintainers decided to move to another mod-loader for whatever reason.
4. NeoForge introduced public API breaking changes, which were done multiple times to clean legacy decisions, so a multi-platform setup will be very useful for porting to newer MC versions.

This was already discussed quite a few times in Discord and internally with the Epic Fight team.

</p>
</details> 

<details><summary>Why not StoneCutter?</summary>
<p>

Gradle projects don't change the Java packages or how they are imported. It's just a powerful way to separate builds, scripts, and dependencies.

StoneCutter, while it's great for multi-version, it's heavily dependent on preprocessor comments:

<img width="989" height="222" alt="image" src="https://github.com/user-attachments/assets/4eb7cd44-4796-41ed-a588-354f17131846" />

So it's less modular, as all platform implementations are in the same Gradle project (`src/main`), and a specific platform cannot be removed easily by simply removing the corresponding project directory.

It's harder to share the platform JAR files for third parties, and therefore, it's less efficient for other multi-platform mods to integrate with Epic Fight, unless they use StoneCutter too.

StoneCutter can still be used for MC multi-version, but the setup is complex, and the build performance is not ideal for convenience.

Also, it does not solve the issue of third-party dependencies.

</p>
</details> 

### Non-goals

- Continue supporting Forge, or explicit Quilt support.
- Complete Fabric port in `1.20.1` or `1.21.1`.
- Supporting alternative mod platforms, such as Liteloader.
  - This step will make it easier for third parties to integrate; however, it's not officially planned or something we are looking forward to, due to the amount of work and considerations required.
- Discounting the NeoForge support.

### Other Questions

<details><summary>What is libs.versions.toml?</summary>
<p>

The versions are stored in Gradle version catalogs (`libs.versions.toml`) instead of `gradle.properties` (the legacy approach), which is more suitable for multi-project sharing of Gradle coordinates, versions, and plugins in a type-safe convivnent way.

<img width="3238" height="2174" alt="image" src="https://github.com/user-attachments/assets/6dd9e5f2-eb98-4a63-a606-61d01a661a2e" />

This also makes the `dependencies` in `common`, `neoforge`, and `fabric` more readable and minimal, allowing for changes to versions, Maven repositories, or coordinates from one central place.

</p>
</details> 

<details><summary>What is strictMaven() function?</summary>
<p>

Preferred over the regular `maven { ... }` as it filters to specific groups and not all of them to reduce network calls, and therefore faster builds.
Typically, mod Maven repositories include several groups and dependencies that are not needed, so it's better not to look them up.

```diff
- maven {
-    name = "NeoForged"
-    url = uri("https://maven.neoforged.net/releases")
- }
+ strictMaven(
+    name = "NeoForged",
+    url = "https://maven.neoforged.net/releases",
+    includeGroups = "net.neoforged",
+ )
```

</p>
</details>

EchoEllet committed:
build: initial multi-platform setup via Gradle projects (common, fabric and neoforge)

Forixaim reviewed (approved):
LGTM.

sentwayfarer commented on fabric/src/main/resources/fabric.mod.json:
It seems that the `authorsList` doesn't have quotes around it like the rest of them. Syntax error?

EchoEllet replied:
Good question. 

This will be replaced with a valid JSON list by Gradle scripts by processing the`authors` field in `gradle.properties` (names separated by a comma).

```kotlin
// gradle/build-logic/conventions/src/main/kotlin/SharedUtils.kt
"authorsList" to modAuthors
            .split(",").joinToString(
                prefix = "[\"",
                postfix = "\"]",
                separator = "\", \""
            ) { it.trim() },
```

The IDE or any tooling thinks this is an invalid JSON file, but it's totally valid.

Run `./gradlew fabric:build` and then take a look at the processed JSON file in `fabric/build/resources/main/fabric.mod.json`.

That's the change:

```diff
- "authors": ${authorsList},
+ "authors": ["Yesman(a.k.a maninthe_home)", "Gui", "Asan", "Squoshi", "Fori", "Ellet"],
```

Yes, I agree that we need to do something about the fact that the IDE and other tools think it's invalid JSON, and it's confusing, but the only workaround I see is:

```json
"authors": [
    "${authors}"
  ],
```

<details><summary>Details</summary>
<p>

When all authors are in the same `String`:

<img width="1712" height="1020" alt="image" src="https://github.com/user-attachments/assets/5b5c0467-325e-4492-9931-baf9a03c5ac0" />

When each author is in a separate `String` in the JSON list:

<img width="846" height="507" alt="image" src="https://github.com/user-attachments/assets/591106aa-52bd-4a00-b1f0-b3b5ec9afaab" />




</p>
</details> 

Maybe this is better:

```json
"authors": [ ${authorsList} ],
```

EchoEllet replied:
> Maybe this is better:

I have [changed it in this push](https://github.com/Epic-Fight/epicfight/compare/8acd12e97e7136d9011135c195ed69ebcfb485e6..9dfc5e663d359c516a8617a55fd511525c9ed20f) with a clearer name, and now the `[]` is part of the source JSON file.

sentwayfarer reviewed (changes requested):
One change needed.

sentwayfarer reviewed (approved):
LGTM

EchoEllet committed:
chore: move NeoForge generated data files from src/generated -> neoforge/src/generated

EchoEllet committed:
chore: move NeoForge loot modifiers from root project to neoforge Gradle project

EchoEllet committed:
refactor: move vanilla MC assets from root -> common Gradle project to share with Fabric

EchoEllet committed:
chore: move Java files from root -> neoforge Gradle project

These files will be slowly moved to common Gradle project, and then implemented in Fabric slowly

EchoEllet committed:
chore: move the remaining asset files from the root to neoforge Gradle project

These should be extracted and moved to the common later, to share with fabric

EchoEllet commented:
@Yesssssman
I have rebased the branch to address all conflicts with the recent changes.

<details><summary>The CI failure is irrelevant to the changes</summary>
<p>

<img width="1182" height="431" alt="image" src="https://github.com/user-attachments/assets/ea498336-0488-4081-bced-92d57f2b4cdc" />

It seems that there is an unchecked file that is used in a checked file.

</p>
</details>

Yesssssman reviewed (approved):
Looks cool, still I'm trying to understand multi loader shape by myself through local dummy project. That way I can figure out some details of this structure.
